### PR TITLE
Adding file lock on global results file

### PR DESCRIPTION
**Pull request by sebhrusen**
addressing https://github.com/openml/automlbenchmark/issues/135
applied on top of improvements regarding saving logic in https://github.com/openml/automlbenchmark/pull/452

### CODE REVIEW

File: frameworks/shared/requirements.in [hunk 1]
```diff
@@ -1,5 +1,4 @@
 psutil>=5.4
-ruamel.yaml>=0.15
```
Comment (sebhrusen) [inline frameworks/shared/requirements.in:2]: as I was rebuilding the dependencies, I also refactored imports in `amlb.utils` to be able to remove this one from frameworks' requirements.

**Comment by sebhrusen** [comment]
@PGijsbers another one for you when you find some review time.

File: amlb/utils/config.py [hunk 2]
```diff
@@ -1,42 +1,49 @@
...
+    def yaml_load(file, as_namespace=False):
+        if as_namespace:
+            yaml = YAML(typ='safe', pure=True)
+            yaml.Constructor = _YAMLNamespaceConstructor
+        else:
+            yaml = YAML(typ='safe')
+        return yaml.load(file)
+else:
+    def yaml_load(*_, **__):
+        raise ImportError("ruamel.yaml package is required to load `yaml` config files.")
```
Comment (PGijsbers) [inline amlb/utils/config.py:46]: Why do we have this construction as opposed to e.g., importing the modules in functions? Wouldn't that accomplish the same?

File: frameworks/shared/caller.py [hunk 3]
```diff
@@ -123,7 +123,7 @@ def run_in_venv(caller_file, script_file: str, *args,
             PYTHONPATH=os.pathsep.join([
                 rconfig().root_dir,
             ]),
-            AMLB_PATH=os.path.join(rconfig().root_dir, "amlb"),
+            AMLB_PATH=os.path.join(rconfig().root_dir),
```
Comment (PGijsbers) [inline frameworks/shared/caller.py:126]: What changed here?

**Review by PGijsbers — COMMENTED** [review]
Thanks for the fix (and the renaming of some functions/properties to better reflect their use is also appreciated) 🎉 

I appreciate the effort for removing the ruamel dependency, but would have appreciated it more as a separate PR. I realise I'm probably wrong but my gut reaction is that this seems too "complicated" for accomplishing the simple task of being able to import something when not all dependencies are installed. I'd like to have a closer look at this aspect, but am on board with the changes proposed that actually address #135.

File: frameworks/shared/caller.py (hunk 3 above)
Comment (sebhrusen) [reply frameworks/shared/caller.py:126]: I changed the semantic of this variable for https://github.com/openml/automlbenchmark/pull/453/files#diff-9a256fecc9fc69ecce8bcb888ebab5292a2b17233713e2c883cd97d36b86f166

user to point to `automlbenchmark/amlb`, now points to root (`automlbenchmark`)

**Pull request merged**